EssentialsDiscordLink/src/main/java/net/essentialsx/discordlink/listeners/LinkBukkitListener.java

```diff
@@ -0,0 +1,158 @@
+package net.essentialsx.discordlink.listeners;
+
+import net.essentialsx.api.v2.events.AsyncUserDataLoadEvent;
+import net.essentialsx.api.v2.events.discord.DiscordMessageEvent;
+import net.essentialsx.discordlink.DiscordLinkSettings;
+import net.essentialsx.discordlink.EssentialsDiscordLink;
+import net.essentialsx.discordlink.UserLinkStatusChangeEvent;
+import org.bukkit.Bukkit;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.EventPriority;
+import org.bukkit.event.Listener;
+import org.bukkit.event.player.AsyncPlayerChatEvent;
+import org.bukkit.event.player.AsyncPlayerPreLoginEvent;
+import org.bukkit.event.player.PlayerCommandPreprocessEvent;
+import org.bukkit.event.player.PlayerInteractEvent;
+
+import static com.earth2me.essentials.I18n.tl;
+
+public class LinkBukkitListener implements Listener {
+    private final EssentialsDiscordLink ess;
+
+    public LinkBukkitListener(EssentialsDiscordLink ess) {
+        this.ess = ess;
+    }
+
+    @EventHandler(priority = EventPriority.HIGH)
+    public void onConnect(final AsyncPlayerPreLoginEvent event) {
+        if (ess.getSettings().getLinkPolicy() != DiscordLinkSettings.LinkPolicy.KICK) {
+            return;
+        }
+
+        if (!ess.getLinkManager().isLinked(event.getUniqueId())) {
+            String code;
+            try {
+                code = ess.getLinkManager().createCode(event.getUniqueId());
+            } catch (IllegalArgumentException e) {
+                code = e.getMessage();
+            }
+            event.disallow(AsyncPlayerPreLoginEvent.Result.KICK_OTHER, tl("discordLinkLoginKick", "/link " + code, ess.getSettings().getInviteUrl()));
+        }
+    }
+
+    @EventHandler(priority = EventPriority.LOW)
+    public void onInteract(final PlayerInteractEvent event) {
+        if (ess.getSettings().getLinkPolicy() != DiscordLinkSettings.LinkPolicy.FREEZE) {
+            return;
+        }
+
+        if (!ess.getLinkManager().isLinked(event.getPlayer().getUniqueId())) {
+            event.setCancelled(true);
+        }
+    }
+
+    @EventHandler(priority = EventPriority.LOW)
+    public void onCommand(final PlayerCommandPreprocessEvent event) {
+        if (ess.getSettings().getLinkPolicy() != DiscordLinkSettings.LinkPolicy.FREEZE) {
+            return;
+        }
+
+        //todo maybe allowed commands
```

Comment:
Might be a good idea, so servers can let people run `/rules` etc

Reply:
Alternatively provide a LuckPerms context for this?

Reply:
> Alternatively provide a LuckPerms context for this?

linked is already a context